bundles/mapping/mapmodule/service/map.layer.js

```diff
@@ -329,10 +346,222 @@ Oskari.clazz.define('Oskari.mapframework.service.MapLayerService',
+                    group = me._layerGroups[i];
+
+                    // Check if layer is in main groups
+                    layerIndex = isLayerInGroup(group.layers);
+                    if (layerIndex >= 0) {
+                        if (deleteLayer) {
+                            group.layers.splice(layerIndex, 1);
+                        } else {
+                            group.layers[layerIndex] = newLayerConf;
+                        }
+                    }
+
+                    // check subgroups
+                    for (var j = 0; j < group.groups.length; j++) {
+                        var subgroup = group.groups[j];
+
+                        layerIndex = isLayerInGroup(subgroup.layers);
+                        if (layerIndex >= 0) {
+                            if (deleteLayer) {
+                                group.groups[j].layers.splice(layerIndex, 1);
+                            } else {
+                                group.groups[j].layers[layerIndex] = newLayerConf;
+                            }
+                        }
+                        // check subgroup subgroups
+                        if (subgroup.groups) {
+                            for (var k = 0; k < subgroup.groups.length; k++) {
+                                var subgroupSubgroup = subgroup.groups[k];
+                                layerIndex = isLayerInGroup(subgroupSubgroup.layers);
+                                if (layerIndex >= 0) {
+                                    if (deleteLayer) {
+                                        group.groups[j].groups[k].layers.splice(layerIndex, 1);
+                                    } else {
+                                        group.groups[j].groups[k].layers[layerIndex] = newLayerConf;
+                                    }
+                                }
+                            }
+                        }
+
+                        if (layerIndex >= 0) {
+                            break;
+                        }
+                    }
+
+                    if (layerIndex >= 0) {
+                        break;
+                    }
+
+                }
+
+
+                // Finally check at layer has all groups in dom
+                newLayerConf.groups.forEach(function(group) {
+                    var groupConf = me.getAllLayerGroups(group.id);
+                    var isInGroup = jQuery.grep(me.getAllLayerGroups(group.id).layers, function(layer) {
+                        return layer.id === newLayerConf.id;
+                    });
+
+                    if (isInGroup.length === 0) {
+                        me.getAllLayerGroups(group.id).layers.push(newLayerConf);
+                    }
+                });
+            }
+
+            // Also check if layer has removed from groups
+            if (deleteLayer) {
+                // find all groups where layer is
+                // check main groups
+                var hasInGroupLayers = function(group) {
```

Comment:
This and the layer index function below have already been implemented in the code above inside the same gigantic function. You can do this with just one generic helper function that takes an array and an id, loops the array to find if any item in the array has a property id with the value of the parameter and return the index or -1 if not found. This can be used for layers and groups as a single implementation instead of having the same thing written 4 times inside the monstrously large function :) Also consider moving atleast delete to it's own function.